File: 0502 IPO/ipo.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        vector<pair<int, int>> projects;
        for (int i = 0; i < profits.size(); i++) {
            projects.push_back(make_pair(capital[i], profits[i]));
        }
        sort(projects.begin(), projects.end());
        priority_queue<int> maxProfits;
        int i = 0;
        while (k > 0) {
            while (i < projects.size() && projects[i].first <= w) {
                maxProfits.push(projects[i].second);
                i++;
            }
            if (maxProfits.empty()) {
                return w;
            }
            int currProfit = maxProfits.top();
            maxProfits.pop();
            w += currProfit;
            k--;
        }
        return w;
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 502-ipo/ipo.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    private static class Project {
        int capital;
        int profit;
        Project(int cap, int proj) {
            capital = cap;
            profit = proj;
        }
    }
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        List<Project> projects = new ArrayList<>();

        for(int i=0; i<n; i++) {
            projects.add(new Project(capital[i], profits[i]) );
        }

        Collections.sort(projects, (a,b)->a.capital-b.capital);

        PriorityQueue<Integer> pq = new PriorityQueue<>( (x, y) -> y-x );
        int i = 0;
        for(int j=0; j<k; j++) {
            while(i<n && projects.get(i).capital<=w) {
                pq.add(projects.get(i).profit);
                i++;
            }
            if(pq.isEmpty()) break;
            w += pq.poll();
        }

        return w;
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
